src/size_hint.rs

```diff
@@ -1,149 +1,379 @@
-//! Utilities for working with and combining the results of
-//! [`Arbitrary::size_hint`][crate::Arbitrary::size_hint].
+//! The [`SizeHint`] type, for use with
+//! [`Arbitrary::size_hint()`].
+
+use core::{fmt, ops};
+
+#[cfg(doc)]
+use crate::Arbitrary;
 
 pub(crate) const MAX_DEPTH: usize = 20;
 
-/// Protects against potential infinite recursion when calculating size hints
-/// due to indirect type recursion.
-///
-/// When the depth is not too deep, calls `f` with `depth + 1` to calculate the
-/// size hint.
-///
-/// Otherwise, returns the default size hint: `(0, None)`.
+/// Bounds on the number of bytes an [`Arbitrary::arbitrary()`] implementation might consume.
 ///
-/// <div class="warning">This method is deprecated. Users should instead implement <a href="../trait.Arbitrary.html#method.try_size_hint"><code>try_size_hint</code></a> and use <a href="fn.try_recursion_guard.html"><code>try_recursion_guard</code></a></div>
-#[inline]
-#[deprecated(note = "use `try_recursion_guard` instead")]
-pub fn recursion_guard(
-    depth: usize,
-    f: impl FnOnce(usize) -> (usize, Option<usize>),
-) -> (usize, Option<usize>) {
-    if depth > MAX_DEPTH {
-        (0, None)
-    } else {
-        f(depth + 1)
+/// A size hint consists of a finite lower bound and a possibly infinite upper bound.
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+pub struct SizeHint {
+    /// Inclusive lower bound.
+    lower: usize,
+
+    /// Exclusive upper bound. [`None`] means unbounded.
+    upper: Upper,
+}
+
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+enum Upper {
+    Bounded(usize),
+    /// The size is known to be unbounded (or overflow [`usize`]).
+    Unbounded,
+    /// The size is unknown because a more specific size hint was not given.
+    /// This is functionally identical to `Unbounded` but prints differently.
+    Unknown,
```

Comment:
I'm not convinced that there is a meaningful difference between these variants. Just having a different `Debug` print doesn't really seem worth it to me.

Reply:
I included this feature precisely because, if it had existed, it would have helped me troubleshoot a missing size hint.